### DISCUSSION TITLE
Geolocation problem 2.1.1 - country is not changing (X-Forwarded-For ignored)

### DISCUSSION THREAD

**Discussion by mobycrypt** [plausible/analytics#4594] [Self-Hosted Support]
My plausible is behind nginx reverse proxy. I set headers to forward to Plausible:
```
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $host;
```

I see in Caddy logs that everything is received as expected, `X-Forwarded-For` header is set on Caddy side with single value: client IP. Geolocation hower is not working correctly, it uses reverse proxy IP rather using `X-Forwarded-For` value. I've tested with curl and it shows different country when I'm on different IP (changed with VPN), ignoring `X-Forwarded-For`. It looks like a bug.

I made temporary **workaround** for that issue. On nginx side I set:
```
    proxy_set_header CF-Connecting-IP $proxy_add_x_forwarded_for;
```
and then plausible uses correct IP.

**Edit.**
It's worth to mention that because country was not changing, my unique visitors metric also looked incorrect (too small). When country works correctly due to  `CF-Connecting-IP` - unique visitors metrics looks good.

**Comment by ruslandoga** [comment]
👋 @PeerZetZzZzZ 

Would you be able to trace the headers Plausible receives? If yes, please search for recon_trace in discussions for how to do it :) That information would speed up finding the root cause of this issue.

**Reply by mobycrypt** [reply]
Here is JSON taken from Caddy docker logs. Sensitive data are replaced (IPs and domains) but in all places so the structure is correct.

```
{
  "level": "info",
  "ts": 1726679629.5630805,
  "logger": "http.log.access.log0",
  "msg": "handled request",
  "request": {
    "remote_ip": "111.111.111.111",
    "remote_port": "48716",
    "client_ip": "111.111.111.111",
    "proto": "HTTP/1.1",
    "method": "POST",
    "host": "domain.where.plausible.is.hosted.com",
    "uri": "/api/event",
    "headers": {
      "Cf-Connecting-Ip": [
        "222.222.222.222"
      ],
      "Sec-Ch-Ua": [
        "\"Not)A;Brand\";v=\"99\", \"Opera GX\";v=\"113\", \"Chromium\";v=\"127\""
      ],
      "Sec-Purpose": [
        "prefetch;prerender"
      ],
      "User-Agent": [
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36 OPR/113.0.0.0"
      ],
      "Purpose": [
        "prefetch"
      ],
      "Sec-Fetch-Site": [
        "same-origin"
      ],
      "Accept-Encoding": [
        "gzip, deflate, br, zstd"
      ],
      "X-Forwarded-For": [
        "222.222.222.222"
      ],
      "Sec-Fetch-Mode": [
        "cors"
      ],
      "Cookie": [
        "REDACTED"
      ],
      "Sec-Ch-Ua-Mobile": [
        "?0"
      ],
      "Sec-Ch-Ua-Platform": [
        "\"macOS\""
      ],
      "Connection": [
        "close"
      ],
      "X-Forwarded-Proto": [
        "https"
      ],
      "X-Forwarded-Host": [
        "application-domain.com"
      ],
      "Content-Length": [
        "78"
      ],
      "Content-Type": [
        "text/plain"
      ],
      "Accept": [
        "*/*"
      ],
      "Origin": [
        "https://application-domain.com"
      ],
      "Sec-Fetch-Dest": [
        "empty"
      ],
      "Accept-Language": [
        "en-GB,en-US;q=0.9,en;q=0.8"
      ],
      "Priority": [
        "u=1, i"
      ],
      "Referer": [
        "https://application-domain.com/"
      ]
    },
    "tls": {
      "resumed": false,
      "version": 772,
      "cipher_suite": 4865,
      "proto": "",
      "server_name": "domain.where.plausible.is.hosted.com"
    }
  },
  "bytes_read": 78,
  "user_id": "",
  "duration": 0.09355184,
  "size": 2,
  "status": 202,
  "resp_headers": {
    "Content-Length": [
      "2"
    ],
    "Content-Type": [
      "text/plain; charset=utf-8"
    ],
    "X-Request-Id": [
      "F_Zl-YN9dV8tNjgAALah"
    ],
    "Server": [
      "Caddy",
      "Cowboy"
    ],
    "Access-Control-Allow-Credentials": [
      "true"
    ],
    "Access-Control-Expose-Headers": [
      ""
    ],
    "Cache-Control": [
      "max-age=0, private, must-revalidate"
    ],
    "Alt-Svc": [
      "h3=\":443\"; ma=2592000"
    ],
    "Access-Control-Allow-Origin": [
      "*"
    ],
    "Date": [
      "Wed, 18 Sep 2024 17:13:49 GMT"
    ]
  }
}

```

**Reply by mobycrypt** [reply]
@ruslandoga let me know if anything extra is needed, I'll try to help

**Reply by ruslandoga** [reply]
We need to know the headers Plausible receives, not Caddy :)

https://github.com/plausible/analytics/discussions/3248#discussioncomment-6670217
https://github.com/plausible/analytics/discussions/2767#discussioncomment-5367417

**Reply by mobycrypt** [reply]
@ruslandoga Here it comes - `X-Forwarded-For` is set with my proxy IP what explains why it resolves country wrong. The question why it has this value when proxy sends client's IP.

```
[
  {
    "accept",
    "*/*"
  },
  {
    "accept-encoding",
    "gzip, deflate, br, zstd"
  },
  {
    "accept-language",
    "pl-PL,pl;q=0.9,en-US;q=0.8,en;q=0.7"
  },
  {
    "cf-connecting-ip",
    "222.222.222.222"
  },
  {
    "content-length",
    "111"
  },
  {
    "content-type",
    "text/plain"
  },
  {
    "host",
    "domain.where.plausible.is.hosted.com"
  },
  {
    "origin",
    "https://application-domain.com"
  },
  {
    "priority",
    "u=1, i"
  },
  {
    "referer",
    "https://application-domain.com"
  },
  {
    "sec-ch-ua",
    "\"Chromium\";v=\"128\", \"Not;A=Brand\";v=\"24\", \"DuckDuckGo\";v=\"128\""
  },
  {
    "sec-ch-ua-mobile",
    "?1"
  },
  {
    "sec-ch-ua-platform",
    "\"Android\""
  },
  {
    "sec-fetch-dest",
    "empty"
  },
  {
    "sec-fetch-mode",
    "cors"
  },
  {
    "sec-fetch-site",
    "same-origin"
  },
  {
    "user-agent",
    "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Mobile Safari/537.36"
  },
  {
    "x-forwarded-for",
    "111.111.111.111"
  },
  {
    "x-forwarded-host",
    "domain.where.plausible.is.hosted.com"
  },
  {
    "x-forwarded-proto",
    "https"
  },
  {
    "x-requested-with",
    "com.duckduckgo.mobile.android"
  }
]
```

**Reply by ruslandoga** [reply]
One interesting thing I see in these headers is that `x-forwarded-for` contains a single value, and it should be at least two since the request passes through at least two proxies (nginx, Caddy). It might be that the last proxy (is it Caddy?) rewrites `x-forwarded-for` instead of extending it.

**Reply by mobycrypt** [reply]
@ruslandoga Here is my nginx proxy/application (docker latest version) config (I cut some locations, but what relevant is there). I serve plausible.js from my application server this is why this location is so short:
```
server {
  server_name application-domain.com www.application-domain.com;
  ssl_certificate /certs/live/data/fullchain.pem;
  ssl_certificate_key /certs/live/data/privkey.pem;

  listen 443 ssl;

  if ($host = www.application-domain.com) {
  return 301 https://application-domain.com$request_uri;
}

http2 on;
gzip off;
gzip_vary on;
gzip_proxied any;
gzip_comp_level 6;
gzip_buffers 16 8k;
gzip_http_version 1.1;
gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript image/svg+xml;
charset utf-8;
proxy_hide_header X-Powered-By;
server_tokens off;
set $plausible_event_url https://domain.where.plausible.is.hosted.com/api/event;

location /images {
try_files $uri $uri/;
}

        
location /plausible.js {
root /usr/share/nginx/html;
try_files $uri $uri/ /plausible.js;
}


location /api/event {
resolver 9.9.9.9 valid=300s;

proxy_pass $plausible_event_url;
proxy_set_header Host domain.where.plausible.is.hosted.com;
proxy_ssl_server_name on;
proxy_ssl_name domain.where.plausible.is.hosted.com;
proxy_buffering on;
proxy_http_version 1.1;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-Host $host;
proxy_set_header CF-Connecting-IP $proxy_add_x_forwarded_for;
}

}
server {
listen 80;
# CHANGE IT
server_name www.application-domain.com application-domain.com;
return 301 https://$host$request_uri;

}

```

I must admit there I experimented with multiple values in `X-Forwarded-For` (like `'X-Forwarded-For <clientIP> <proxyIP>'`) using CURL and it didn't work too. I can try these tests again. Thank you for your support so far.

edit. I noticed that for my experiments i disabled GZIP at all :D

**Reply by ruslandoga** [reply]
What about Caddy and any other reverse proxies in the request's path to Plausible? If Caddy receives expected `x-forwarded-for` from nginx but Plausible receives something else, the problem might be in Caddy or between Caddy and Plausible.

**Reply by mobycrypt** [reply]
@ruslandoga The first JSON I provided in this thread is taken directly from Caddy docker logs. Client IP is there in `X-Forwarded-For` when it comes to Caddy. No other proxies between, Caddy receives directly from nginx proxy.

**Reply by ruslandoga** [reply]
I meant, how is it configured?

**Reply by mobycrypt** [reply]
Caddy configuration is default setup based on this: https://github.com/plausible/community-edition. The only plausible `labels` set in docker-compose are: `virtual.host, virtual.port, virtual.tls-email`

**Reply by ruslandoga** [reply]
I see. So the problem seems to be in Caddy.

From https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#defaults:

> For these X-Forwarded-* headers, by default, the proxy will ignore their values from incoming requests, to prevent spoofing.

**Reply by ruslandoga** [reply]
Note that since you are already using nginx, Caddy is optional and nginx can send requests directly to Plausible. That's one way to fix the issue. Another is to add extra configuration to `caddy-gen` service, it seems to support relevant directives after https://github.com/wemake-services/caddy-gen/pull/52

**Reply by ruslandoga** [reply]
Ah, and your `proxy_set_header CF-Connecting-IP $proxy_add_x_forwarded_for;` workaround seems fine as well, at least for now. The only potential problem with it is that it relies on the current header priority in [`Plausible.RemoteIP`](https://github.com/plausible/analytics/blob/5766478887929a17c5c03105a9381d68968abf25/lib/plausible_web/remote_ip.ex#L6-L36) where `cf-connecting-ip` is preferred over `x-forwarded-for`. It might stop working if this Plausible's logic in picking the header to read the client IP from  changes. Setting `x-plausible-ip` might be a bit safer.

**Reply by mobycrypt** [reply]
@ruslandoga I have 2 hosts on different domains: application host + plausible host. So both hosts have own proxies. I replaced `cf-connecting-ip` with `x-plausible-ip` and looks like working fine, thanks!

### So to summarize:

Caddy is ignoring X-Forwarded-* headers this is why country is not correctly obtained based on `X-Forwarded-For` IP address value. Adding `proxy_set_header X-Plausible-IP $proxy_add_x_forwarded_for;`  to nginx proxy is a workaround, alternatively (`CF-Connecting-IP` instead of `X-Plausible-IP`).

Thank you @ruslandoga for your support
